### PR TITLE
Emit `ContractReverted` error when revert flag is set

### LINKED ISSUE #9338: Return `ContractReverted` from dispatchables when top level contract reverted
Contracts have the possibility to set a flag that reverts all the changes they made. The calling contract can evaluate these flags and act on it. However, if the top level contract sets the revert flag when returning the extrinsic is counted as `Ok`. We should introduce a new pallet level error `Err(ContractReverted)` that is returned instead.

Please note that `seal_return` allows to return a result buffer that can be used to encode additional error information. The buffer is only accessible when calling via RPC of course. How this buffer is to be decoded (by a UI) for a specific contract should be defined by the userspace ABI (json). When called via RPC we return `Ok` so that buffer is accessible. UIs should inspect the `flags` field in order to determine whether the call was a success.

**Please note that this should only happen if the call/instantiate was submitted as transaction where the `flags` field cannot be read. An RPC call should still return OK and it is up to the UI to classify it as error by reading `flags`. Reason is that on dispatch error no output buffer could be returned.**

### PR DESCRIPTION
**Pull request by athei**
Fixes #9338

No changes to ink! or UI necessary.

### CODE REVIEW

File: frame/contracts/src/lib.rs [hunk 1]
```diff
@@ -356,6 +361,11 @@ pub mod pallet {
 				salt,
 				None,
 			);
+			if let Ok(retval) = &output.result {
+				if !retval.1.is_success() {
```
Comment (ascjones) [inline frame/contracts/src/lib.rs:365]: Is it plausible that the implementation of `is_success` changes to check for other flags? If so would it make sense to add a specific `is_reverted` method in addition?

Comment (athei) [reply frame/contracts/src/lib.rs]: I think this new method should replace `is_success`. No reason to have it as long as there is no other failure flag. Better be specific.

**Review by ascjones — APPROVED** [review]
LGTM

**Review by cmichi — APPROVED** [review]

**Comment by athei** [comment]
bot merge

**Pull request merged**